SAAS-1831 - app system local development (#170)
shopware/docs: guides/plugins/apps/local-development/app-development-with-docker.md

For further information about the `manifest.xml` have a look at our [documentation](https://docs.shopware.com/en/shopware-platform-dev-en/app-system-guide/setup#manifest-file).  
To easily create a `manifest.xml` take a look at our [create-manifest command](../platform-sh-template.md#create-the-manifestxml).

Your file structure should look as follows:
```
...
│
├──development
│  ├──custom
│  │  └───apps
│  │      └───yourAppName
│  │          └───manifest.xml
│  │
│  ├──platform
│  └──...
│
└──shopwareAppTemplate
...
```

## Combining both in one docker setup

Once your Shopware development setup is ready to go you need to add your app to it. This is done by adding the services to your `development/docker-compose.yml`.  
At first, you need to add two networks. One for your app system and another one for combining the app system with Shopware.
  
This is done by simply adding the networks `appSystem` and `development` to your existing ones:

> ```Yaml
> networks:
>     shopware:
>     appSystem:
>     development:
> ```

Reviewer: ``````suggestion

{% code title="development/docker-composer.yml %}
```Yaml
networks:
    shopware:
    appSystem:
    development:
```
{% endcode %}
``````

The `appSystem`-network is only for your app server and the app database.  
The `development`-network is used to combine your app server with the Shopware server.  

### Adding the app server

Now you need to define the `services` in your `development/docker-compose.yml`. Insert the following to firstly add your app server.   

> ```Yaml
> services:
> [...]
>   example_app_server:
>     image: shopware/development:local
>     volumes:
>       - "../shopwareAppTemplate:/app"
>       - "~/.composer:/.composer"
>     environment:
>       CONTAINER_UID: 1000
>       APPLICATION_UID: 1000
>       APPLICATION_GID: 1000
>     ports:
>       - "127.0.0.1:7777:8000"
>     networks:
>       appSystem:
>       development:
>         aliases:
>           - example
> ```

Reviewer: ``````suggestion
{% code title="development/docker-compose.yml" %}
```Yaml
services:
[...]
  example_app_server:
    image: shopware/development:local
    volumes:
      - "../shopwareAppTemplate:/app"
      - "~/.composer:/.composer"
    environment:
      CONTAINER_UID: 1000
      APPLICATION_UID: 1000
      APPLICATION_GID: 1000
    ports:
      - "127.0.0.1:7777:8000"
    networks:
      appSystem:
      development:
        aliases:
          - example
```
{% endcode %}
``````

> This adds a new container to your docker setup running your app server's code. The new container is available at the networks `appSystem` and `development`.  

Reviewer: ```suggestion
This adds a new container to your docker setup running your app server's code. The new container is available inside the networks `appSystem` and `development`.  
```

In the `development`-network your app server has the alias `example`. This will be the url which your Shopware server needs to communicate with. This is also the url which you should use in your `manifest.xml` except for iframes.  

> The `volumes` represents the relative path to your app.  

Reviewer: ```suggestion
`volumes` represents the relative path to your app.  
```

> And `ports` exposes port `8000` to `127.0.0.1:7777` to us so that we can go to `127.0.0.1:7777` or `localhost:7777` to directly connect to your app server. This will come in handy when we register your own modules to use iframes.  

Reviewer: ```suggestion
And `ports` exposes port `8000` to `127.0.0.1:7777` to us so that we can visit `127.0.0.1:7777` or `localhost:7777` to directly connect to your app server. This will come in handy when we register our own modules to use iframes.  
```

### Adding the app database

The next step is to also add your mysql server to your docker setup. This is as easy as it was for your app server.  
Simply add this to your `development/docker-compose.yml`.

> ```Yaml
> services:
> [...]
>   example_mysql:
>     build: dev-ops/docker/containers/mysql
>     environment:
>       MYSQL_ROOT_PASSWORD: root
>       MYSQL_USER: app
>       MYSQL_PASSWORD: app
>     ports:
>       - "5506:3306"
>     volumes:
>       - ./dev-ops/docker/_volumes/mysql-example:/mysql-data
>     networks:
>       appSystem:
>         aliases:
>           - appmysql
> ```

Reviewer: ``````suggestion
{% code title="development/docker-compose.yml" %}
```Yaml
services:
[...]
  example_mysql:
    build: dev-ops/docker/containers/mysql
    environment:
      MYSQL_ROOT_PASSWORD: root
      MYSQL_USER: app
      MYSQL_PASSWORD: app
    ports:
      - "5506:3306"
    volumes:
      - ./dev-ops/docker/_volumes/mysql-example:/mysql-data
    networks:
      appSystem:
        aliases:
          - appmysql
```
{% endcode %}
``````

As you already know you connect your mysql server to the same network as your app server and give it the alias `appmysql`. Furthermore, you can now connect to your database on port `5506` from outside of the docker container.  

> Last but not least we define the credentials for the mysql server and you are done with this.  

Reviewer: ```suggestion
Last but not least we define the credentials for the mysql server and you are done with setting up the database container.  
```

### Combining both in one network

Now you need to add your Shopware server to your `development`-network and give him an alias as follows:  

> ```Yaml
> services:
> [...]
>   app_server:
>     image: shopware/development:latest
>     networks:
>       shopware:
>         aliases:
>           - docker.vm
>       development:
>         aliases:
>           - shopware
>     extra_hosts:
>       - "docker.vm:127.0.0.1"
>     volumes:
>       - ~/.composer:/.composer
>     tmpfs:
>       - /tmp:mode=1777
> ```

Reviewer: ``````suggestion
{% code title="development/docker-composer.yml" %}
```Yaml
services:
[...]
  app_server:
    image: shopware/development:latest
    networks:
      shopware:
        aliases:
          - docker.vm
      development:
        aliases:
          - shopware
    extra_hosts:
      - "docker.vm:127.0.0.1"
    volumes:
      - ~/.composer:/.composer
    tmpfs:
      - /tmp:mode=1777
```
{% endcode %}
``````

> Now your app server can communicate with the Shopware server and your app server can also communicate with your app database.

Reviewer: ```suggestion
Now your app server can communicate with the Shopware server and your app's database.
```

## Access your app server via ssh

To easily access your app server via ssh, you need to create this script `development/dev-ops/docker/actions/ssh-app-server.sh`.

> ```shell script
> #!/usr/bin/env bash
> 
> TTY: docker exec -i --env COLUMNS=`tput cols` --env LINES=`tput lines` -u __USERKEY__ -t __EXAMPLE_APP_SERVER_ID__ bash
> ```

Reviewer: ``````suggestion
{% code title="development/dev-ops/docker/actions/ssh-app-server.sh" %}
```shell script
#!/usr/bin/env bash

TTY: docker exec -i --env COLUMNS=`tput cols` --env LINES=`tput lines` -u __USERKEY__ -t __EXAMPLE_APP_SERVER_ID__ bash
```
{% endcode %}
``````

This script can be executed from your `development` folder with `./psh.phar docker:ssh-app-server`. Keep in mind that this is only possible when the app server has been started with `./psh.phar docker:start` from your development folder.  

To make sure this script actually knows the ID of your app server which is running in the docker container, you need to define the `EXAMPLE_APP_SERVER_ID` in the `development/.psh.yaml.override`.  
Your `development/.psh.yaml.override` should look like this:

> ```Yaml
> ...
> dynamic:
>   ...
>   EXAMPLE_APP_SERVER_ID: docker-compose ps -q example_app_server
>   ...
> ...
> ```

Reviewer: ``````suggestion
{% code title="development/.psh.yaml.override" %}
```Yaml
# ...
dynamic:
 #  ...
  EXAMPLE_APP_SERVER_ID: docker-compose ps -q example_app_server
 #  ...
# ...
```
{% endcode %}
``````

## Initialising the app server
To initialise the app server and the app database you should access your app server via ssh and run `composer install --no-interaction`.  
Next you need to change your `shopwareAppTemplate/.env` and set the DATABASE_URL to `mysql://app:app@appmysql:3306/main`.  
This should look familiar to you because you just configured it in the `development/docker-compose.yml`.  
  

> The next steps should be done while connected over ssh to your app server.  

Reviewer: ```suggestion
The next steps should be done while being connected over ssh to your app server.  
```

> Now you can set up the database by simply typing `bin/console doctrine:database:create`. This will create your database with name `main`.  

Reviewer: ```suggestion
Now you can set up the database by simply executing `bin/console doctrine:database:create`. This will create your database with the name `main`.  
```

Then execute the migrations with `bin/console doctrine:migrations:migrate --no-interaction`. Now your database is ready.  

## Registration

This last step assumes that you already have a valid `manifest.xml` in the correct folder. In order to check this, make sure your `manifest` is in `development/custom/apps/yourAppName/manifest.xml`.  
Then access your local Shopware instance via ssh with `./psh.phar docker:ssh` and execute the check with `bin/console app:validate`. This will tell you if you provided a valid `manifest.xml`.

For the sake of simplicity you need to change the `APP_URL` of your Shopware instance to match the network-alias you gave him.  
This should be done in your `development/.psh.yaml.override` which should look like this:

> ```Yaml
> ...
> const:
>   ...
>   APP_URL: "http://shopware"
>   ...
> ...
> ```  

Reviewer: ``````suggestion
{% code title="development/.psh.yaml.override" %}
```Yaml
# ...
const:
  # ...
  APP_URL: "http://shopware"
  # ...
# ...
```
{% endcode %}  
``````

# App development with docker

## Overview

This guide will walk you through the process of combining Shopware and your app in one setup.

## File Structure

> At first, you should clone the app template from [GitHub](https://github.com/shopwareLabs/AppTemplate) and create a `manifest.xml` for your app. Or take a glance at our fully working example based on this template: [appExample](https://github.com/shopwareLabs/AppExample).  

Reviewer 2: ```suggestion
At first, you need to clone the app template from [GitHub](https://github.com/shopwareLabs/AppTemplate) and create a `manifest.xml` for your app. Or take a glance at our fully working example based on this template: [appExample](https://github.com/shopwareLabs/AppExample).  
```
This is a manual, as such it needs to be assertative. Words like "should" should be avoided, unless there are legitimate alternatives or it describes the result of a third party systems. E.g "change your dns record and your website **should** be reachable from that domain"

For further information about the `manifest.xml` have a look at our \[PLACEHOLDER-LINK: manifest-documentation\].  

> To easily create a `manifest.xml` take a look at our [create-manifest command](../platform-sh-template.md#create-the-manifestxml).

Reviewer 2: ```suggestion
The easiest way to create a `manifest.xml` take a look at our [create-manifest command](../platform-sh-template.md#create-the-manifestxml).
```

Your file structure should look as follows:
```
...
│
├──development
│  ├──custom
│  │  └───apps
│  │      └───yourAppName
│  │          └───manifest.xml
│  │
│  ├──platform
│  └──...
│
└──shopwareAppTemplate
...
```

## Combining both in one docker setup

Once your Shopware development setup is ready to go you need to add your app to it. This is done by adding the services to your `development/docker-compose.yml`.  
At first, you need to add two networks. One for your app system and another one for combining the app system with Shopware.
  

> This is done by simply adding the networks `appSystem` and `development` to your existing ones:

Reviewer 2: ```suggestion
This is done by adding the networks `appSystem` and `development` to your existing ones:
```

{% code title="development/docker-composer.yml %}
```Yaml
networks:
    shopware:
    appSystem:
    development:
```
{% endcode %}

The `appSystem`-network is only for your app server and the app database.  
The `development`-network is used to combine your app server with the Shopware server.  

### Adding the app server

> Now you need to define the `services` in your `development/docker-compose.yml`. Insert the following to firstly add your app server.

Reviewer 2: ```suggestion
Now you need to define the `services` in your `development/docker-compose.yml`. First insert the following to add your app server.
```

  
The next steps should be done while being connected over ssh to your app server.  
Now you can set up the database by simply executing `bin/console doctrine:database:create`. This will create your database with the name `main`.  
Then execute the migrations with `bin/console doctrine:migrations:migrate --no-interaction`. Now your database is ready.  

## Registration

This last step assumes that you already have a valid `manifest.xml` in the correct folder. In order to check this, make sure your `manifest` is in `development/custom/apps/yourAppName/manifest.xml`.  
Then access your local Shopware instance via ssh with `./psh.phar docker:ssh` and execute the check with `bin/console app:validate`. This will tell you if you provided a valid `manifest.xml`.

For the sake of simplicity you need to change the `APP_URL` of your Shopware instance to match the network-alias you gave him.  
This should be done in your `development/.psh.yaml.override` which should look like this:

{% code title="development/.psh.yaml.override" %}
```Yaml
# ...
const:
  # ...
  APP_URL: "http://shopware"
  # ...
# ...
```
{% endcode %}  

> To make sure your `APP_URL` changed you need to reconnect to your Shopware instance via ssh with `./psh.phar docker:ssh`. Now your `APP_URL` changed and you can register your app via `bin/console app:refresh`.  This can also be done by `bin/console app:install yourAppName`. In order to access the Shopware admin you should run `./psh.phar administration:watch`. Now you can access the admin on `localhost:8080`. 

Author: You're totally right. I will change this to `localhost:8000/admin`
